onnxruntime/core/mlas/lib/qgemm.cpp

```diff
@@ -1147,6 +1152,434 @@ const MLAS_GEMM_U8X8_DISPATCH MlasGemmU8X8DispatchSse = {
 
 #endif
 
+// N.B. MSVC does not require turning on SSE 4.1 intrinsics and the current use
+// for this code is Windows only, so restrict this kernel to that environment.
+#if defined(MLAS_SSE2_INTRINSICS) && defined(_MSC_VER)
+
+struct MLAS_GEMM_U8S8_KERNEL_SSE41
+{
+    typedef uint8_t PackedAType;
+    typedef uint8_t PackedBType;
+    typedef int8_t OffsetBType;
+
+    static constexpr size_t PackedK = 4;
+    static constexpr MLAS_GEMM_U8X8_STRIDES Strides{24, 128, 128};
+    static constexpr MLAS_GEMM_U8X8_STRIDES PackedStrides{24, 128, 128};
+};
+
+constexpr size_t MLAS_GEMM_U8S8_KERNEL_SSE41::PackedK;
+constexpr MLAS_GEMM_U8X8_STRIDES MLAS_GEMM_U8S8_KERNEL_SSE41::Strides;
+constexpr MLAS_GEMM_U8X8_STRIDES MLAS_GEMM_U8S8_KERNEL_SSE41::PackedStrides;
+
+template<>
+void
+MlasGemmU8X8CopyPackA<MLAS_GEMM_U8S8_KERNEL_SSE41>(
+    MLAS_GEMM_U8S8_KERNEL_SSE41::PackedAType* D,
+    const uint8_t* A,
+    size_t lda,
+    size_t CountM,
+    size_t CountK,
+    int32_t* RowSumBuffer
+    )
+{
+    const __m128i ZeroVector = _mm_setzero_si128();
+    const __m128i OnesWordBroadcast = _mm_set1_epi16(1);
+
+    //
+    // Process a single row of matrix A in a loop.
+    //
+
+    while (CountM > 0) {
+
+        const uint8_t* a = A;
+        size_t k = CountK;
+        __m128i ReductionVector = ZeroVector;
+
+        //
+        // Copy the source bytes to the packed buffer.
+        //
+        // The packed buffer has the same data ordering as the source bytes,
+        // but CountK is aligned up to a multiple of 4 to maintain 32-bit
+        // alignment. All extra bytes are zero-padded.
+        //
+
+        while (k >= 8) {
+
+            __m128i Bytes = _mm_loadl_epi64((const __m128i*)&a[0]);
+
+            ReductionVector = _mm_add_epi16(ReductionVector, _mm_unpacklo_epi8(Bytes, ZeroVector));
```

Comment:
Maybe a comment about count K <= 128 here? Or would using 32b accumulator hurt performance here?

Reply:
CountK could have been larger here because the values were spread across the vector, but using a 32-bit accumulator wasn't a noticeable hit, so I updated the code to do that instead.